msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/CloudShellManagedIdentitySource.java

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.aad.msal4j;
+
+import com.nimbusds.oauth2.sdk.util.URLUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.HashMap;
+
+class CloudShellManagedIdentitySource extends AbstractManagedIdentitySource{
+
+    private static final Logger LOG = LoggerFactory.getLogger(CloudShellManagedIdentitySource.class);
+
+    // MSI Constants. Docs for MSI are available here https://learn.microsoft.com/en-us/azure/cloud-shell/msi-authorization
+    private static URI endpointUri;
+
+    private URI endpoint;
+
+    @Override
+    public void createManagedIdentityRequest(String resource) {
+        managedIdentityRequest.baseEndpoint = endpoint;
+        managedIdentityRequest.method = HttpMethod.POST;
+
+        managedIdentityRequest.headers = new HashMap<>();
+        managedIdentityRequest.headers.put("ContentType", "application/x-www-form-urlencoded");
+        managedIdentityRequest.headers.put("Metadata", "true");
+
+        managedIdentityRequest.bodyParameters = new HashMap<>();
+        managedIdentityRequest.bodyParameters.put("resource", Collections.singletonList(resource));
+    }
+
+    private CloudShellManagedIdentitySource(MsalRequest msalRequest, ServiceBundle serviceBundle, URI endpoint)
+    {
+        super(msalRequest, serviceBundle, ManagedIdentitySourceType.CloudShell);
+        this.endpoint = endpoint;
+
+        ManagedIdentityIdType idType =
+                ((ManagedIdentityApplication) msalRequest.application()).getManagedIdentityId().getIdType();
```

Comment:
Can you avoid the upcast?

Reply:
I don't think this can be avoided as the request object has a common interface and the Application object references the interface. The ManagedIdentityId object is specific to managedIdentityApplication.